src/Todo/DB.hs

```diff
@@ -0,0 +1,54 @@
+module Todo.DB
+    ( TodoDB (..)
+    , DBError (..)
+    , addTodo
+    , addTag
+    , deleteTag
+    , editTag
+    ) where
+
+import Data.HashSet (HashSet)
+import Data.Foldable (toList)
+import qualified Data.HashSet as HSet
+
+import Todo.Model (Todo (..), Tag (..))
+import Data.NESet (NonEmptySet (..))
+import qualified Data.NESet as NESet
+
+data TodoDB = TodoDB
+    { dbTodos :: HashSet Todo
+    , dbTags  :: HashSet Tag
+    } deriving (Eq, Show)
+
+data DBError = SingleTaggedTodo | NotRegisteredTags deriving (Eq, Show)
+
+addTodo :: Todo -> TodoDB -> Either DBError TodoDB
+addTodo todo@Todo {..} TodoDB {..} 
+    | f == mempty = Right TodoDB { dbTodos = HSet.insert todo dbTodos, .. }
```

Comment:
I think `HSet.null f` is slightly better than comparison with `mempty`. Also, variable name `f` is not good since your code is not general enough to have general names. Ans usually `f` is for **f**unctions. I would name it as `missingTags` if you don't mind to have long names. Or just `miss` if you like short names.